File: python/open_guardrail/guards/mac_address_detect.py
"""Detect MAC addresses."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_MAC_PATTERN = re.compile(
    r"\b(?:[0-9A-Fa-f]{2}[:-]){5}"
    r"[0-9A-Fa-f]{2}\b"
)


class _MacAddressDetect:
    def __init__(
        self, *, action: str = "block"
    ) -> None:
        self.name = "mac-address-detect"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matches = _MAC_PATTERN.findall(text)
        triggered = len(matches) > 0
        score = (
            min(len(matches) / 2, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="mac-address-detect",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "MAC address detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"mac_count": len(matches)}
                if triggered
                else None
            ),
        )


def mac_address_detect(
    *, action: str = "block"
) -> _MacAddressDetect:
    return _MacAddressDetect(action=action)

File: packages/guards/src/mac-address-detect.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface MacAddressDetectOptions {
  action: 'block' | 'warn';
}

const MAC_PATTERN = /\b(?:[0-9A-Fa-f]{2}[:-]){5}[0-9A-Fa-f]{2}\b/g;

export function macAddressDetect(options: MacAddressDetectOptions): Guard {
  return {
    name: 'mac-address-detect',
    version: '0.1.0',
    description: 'Detects MAC addresses',
    category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matches = text.match(MAC_PATTERN) ?? [];
      const triggered = matches.length > 0;

      return {
        guardName: 'mac-address-detect',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(matches.length / 2, 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { macCount: matches.length } : undefined,
      };
    },
  };
}
